### PR TITLE
Implement C++-side basics of "dynamic shapes" API

### PR DESCRIPTION
**Pull request by ansley**

### CODE REVIEW

File: lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp [hunk 1]
```diff
@@ -408,6 +410,7 @@ std::vector<at::Tensor> LtcCreateTensorList(const at::TensorList& tensors) {
 using namespace torch::lazy;
 
 void InitLtcModuleBindings(py::module m) {
+  py::class_<torch::lazy::Node, std::shared_ptr<torch::lazy::Node>>(m, "IrNode");
```
Comment (Krovatkin) [inline lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp:413]: I don't think we need this for C++ only side?

File: torch/csrc/lazy/core/dynamic_ir.h [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+};
+
+class TORCH_API SizeDiv: public DimensionNode {
+ public:
+  SizeDiv(Value a, Value b);
+  int64_t getStaticValue() const override;
+  std::string ToString() const override;
+};
+
+class TORCH_API SizeInt: public DimensionNode {
```
Comment (Krovatkin) [inline torch/csrc/lazy/core/dynamic_ir.h:109]: I'm not sure we need this. How are we going to use this class?

File: torch/csrc/lazy/core/dynamic_ir.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+std::string DimensionNode::ToString() const {
+  return "DimensionNode";
+}
+
+SizeNode::SizeNode(Value input, size_t dim): 
+    DimensionNode(OpKind{c10::Symbol::fromQualString("aten::size")}, {input}, MHash(dim)),
+    dim_(dim) {};
+
+int64_t SizeNode:: getStaticValue() const {
+    return dynamic_cast<const TsNode*>(operand(0).node)->shape(0).size(dim_);
```
Comment (Krovatkin) [inline torch/csrc/lazy/core/dynamic_ir.cpp:31]: I have a feeling this wont work since you are extending torch::lazy::DimensionNode` from `torch::lazy::Node` rather than `TsNode`. I think it's okay to extend from `TsNode` for now or you could wait until Will lands this PR : https://github.com/pytorch/pytorch/pull/74071 . You might be able to land yr PR first.

File: torch/csrc/lazy/core/dynamic_ir.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <torch/csrc/lazy/core/dynamic_ir.h>
+
+namespace torch {
+namespace lazy {
+
+DimensionNode::DimensionNode(OpKind op, OpList operands, hash_t hash_seed):
+  Node(op, /*num_outputs=*/1,
```
Comment (Krovatkin) [inline torch/csrc/lazy/core/dynamic_ir.cpp:9]: I'm surprised `Node` class is bare-bones. @wconstab do you know why we don't even keep `operands_` in Node?

**Review by Krovatkin — COMMENTED** [review]
Could add some basic tests e.g.

1) construct a lazy tensor, get its IR value and construct a size node on dim 0, call getStaticValue(), make sure it's equal to the lazy tensor .shape().at(0)
2. construct a lazy tensor, get its IR value and construct two size nodes on dim 0 and dim1, create an SizeAdd node using the two size nodes(), call getStaticValue() and make sure it gives the right answer.

File: torch/csrc/autograd/python_variable.cpp [hunk 5]
```diff
@@ -164,7 +164,7 @@ static PyObject* device_to_py_class_ [static_cast<size_t>(c10::DeviceType::COMPI
 void registerPythonTensorClass(const std::string& device, PyObject* python_tensor_class) {
   c10::Device dev(device);
 
-  TORCH_CHECK(dev.type() == kXLA, "Only the python class for XLA can be overriden");
+  TORCH_CHECK(dev.type() == kXLA || dev.type() == kLazy, "Only the python class for XLA and Lazy can be overriden");
```
Comment (Krovatkin) [inline torch/csrc/autograd/python_variable.cpp:167]: we can remove this for now since its' only needed for Python

File: torch/csrc/autograd/python_variable.cpp [hunk 6]
```diff
@@ -222,7 +222,7 @@ PyObject * THPVariable_Wrap(at::TensorBase var)
     }
   }
 
-  if (C10_LIKELY(var.device().type() != c10::kXLA)) {
+  if (C10_LIKELY(var.device().type() != c10::kXLA && var.device().type() != c10::kLazy)) {
```
Comment (Krovatkin) [inline torch/csrc/autograd/python_variable.cpp:225]: we can remove this part for now since it's also needed if we were to use Python subclass extending torch.tensor.

File: torch/csrc/lazy/core/dynamic_ir.h (hunk 2 above)
Comment (ansley) [reply torch/csrc/lazy/core/dynamic_ir.h:109]: It was in [this](https://docs.google.com/document/d/1wO63_l4wQwMRN4s82TYxPX7vGgw-PdN_JWhQ4wUpqCU) doc when I originally added the class, but I agree that it doesn't seem necessary

File: torch/csrc/lazy/core/dynamic_ir.cpp (hunk 3 above)
Comment (wconstab) [reply torch/csrc/lazy/core/dynamic_ir.cpp:31]: i landed #74071 FYI

File: torch/csrc/lazy/core/dynamic_ir.cpp (hunk 4 above)
Comment (wconstab) [reply torch/csrc/lazy/core/dynamic_ir.cpp:9]: i think we could move it back, at one point i thought we might not want it/need it for TS backend, but that didn't pan out

File: torch/csrc/lazy/core/dynamic_ir.h (hunk 2 above)
Comment (Krovatkin) [reply torch/csrc/lazy/core/dynamic_ir.h:109]: 👍

File: lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp [hunk 7]
```diff
@@ -2,7 +2,9 @@
 #include <c10/util/Optional.h>
 #include <torch/csrc/lazy/backend/backend_device.h>
 #include <torch/csrc/lazy/backend/backend_interface.h>
+#include <torch/csrc/lazy/core/dynamic_ir.h>
```
Comment (Krovatkin) [inline lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp:5]: could we remove the entire `init_python_bindings.cpp` from this commit ?

**Review by Krovatkin — APPROVED** [review]
@ansley 

Looks good! Could we 

* remove the entire init_python_bindings.cpp from this commit
* Rebase on Will's latest changes?

File: torch/csrc/lazy/core/ir.cpp [hunk 8]
```diff
@@ -6,6 +6,19 @@ C10_DEFINE_bool(ltc_enable_dynamic_shapes, false, "Whether dynamic shape is enab
 namespace torch {
 namespace lazy {
 
+hash_t OperandHashes(const OpList& operands, const hash_t& seed, bool bakeInSizes) {
```
Comment (wconstab) [inline torch/csrc/lazy/core/ir.cpp:9]: we should align this change with @shunting314's proposed changes to fix what @JackCaoG was asking us to do, to move the hash_with_sizes stuff strictly inside TsNode

**Comment by wconstab** [comment]
This PR will need to be landed to master after #74111, it may or may not need to be landed to staging first (it's probably fine to land to staging as-is)

File: torch/csrc/lazy/core/dynamic_ir.h [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+ * ```
+ * 
+ * Here, `x.shape()[i]` will be a SizeNode (subclass of DimensionNode),
+ * and the multiplication of the two SizeNodes will be represented by
+ * a SizeMul (also a subclass of DimensionNode). Through this, we can
+ * prevent `numel` from being represented as a Python int and thus
+ * burned into the Graph.
+ */
+
+class TORCH_API DimensionNode : public lazy::TsNode {
```
Comment (miladm) [inline torch/csrc/lazy/core/dynamic_ir.h:46]: PT/XLA LTC relies on [`lazy::Node`](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/ir.h#L96). Do you plan to subclass `DimensionNode` from `lazy::Node` later?

File: torch/csrc/lazy/core/dynamic_ir.cpp [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+std::string SizeMul::ToString() const {
+  return "SizeMul";
+}
+
+SizeDiv::SizeDiv(Value a, Value b):
+  DimensionNode(OpKind{c10::Symbol::fromQualString("aten::div")}, {a, b}) {};
+
+int64_t SizeDiv::getStaticValue() const {
+    return dynamic_cast<const DimensionNode*>(operand(0).node)->getStaticValue() / dynamic_cast<const DimensionNode*>(operand(1).node)->getStaticValue();
```
Comment (miladm) [inline torch/csrc/lazy/core/dynamic_ir.cpp:52]: Do we need exception handling against division by zero?

File: torch/csrc/lazy/core/dynamic_ir.h (hunk 9 above)
Comment (miladm) [reply torch/csrc/lazy/core/dynamic_ir.h:46]: Design section [C++ Level API (PyTorch/LTC)](https://docs.google.com/document/d/1qUBwlLbDPsWKP3kaSCEgIF4hf-C4vvpA10upEvZep_s/edit#heading=h.h59w757f5joi)

File: torch/csrc/lazy/core/dynamic_ir.h [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+  std::string ToString() const override;
+};
+
+class TORCH_API SizeDiv: public DimensionNode {
+ public:
+  SizeDiv(Value a, Value b);
+  int64_t getStaticValue() const override;
+  std::string ToString() const override;
+};
+
```
Comment (miladm) [inline torch/csrc/lazy/core/dynamic_ir.h:94]: Great operations.
Is there a plan to include `SizeSub`, `SizeSqrt`, `SizePow`?

**Review by miladm — COMMENTED** [review]
Thanks @ansley!

I added a few comments below.

File: torch/csrc/lazy/core/dynamic_ir.h (hunk 11 above)
Comment (Krovatkin) [reply torch/csrc/lazy/core/dynamic_ir.h:94]: yes

File: torch/csrc/lazy/core/dynamic_ir.cpp (hunk 10 above)
Comment (Krovatkin) [reply torch/csrc/lazy/core/dynamic_ir.cpp:52]: a good point!

File: torch/csrc/lazy/core/dynamic_ir.h (hunk 9 above)
Comment (wconstab) [reply torch/csrc/lazy/core/dynamic_ir.h:46]: @Krovatkin what are the blockers to this exactly?
1) Node was missing shapes_ and operands_ fields, etc
2) we have this general assumption that TsNode is the base of all our nodes

(1) was fixed yesterday by @antoniojkim with #74710
(2) is still an open question for 2 reasons,
  2a - we planned to rely on TsNode to implement shape cache and keep Node simpler/unaware of 'hash_with/without_sizes
  2b - Xla would rely on having XlaNode as a base class at least for a long time, if just for migration purposes

Comment (wconstab) [reply torch/csrc/lazy/core/dynamic_ir.h:46]: in other words, is it going to be OK if DiminsionNode extends from Node?  Or does that violate our design that backends can rely on all nodes deriving from {Backend}Node?

**Comment by miladm** [comment]
Is there any update on the merging plan for this PR? 
@wconstab @Krovatkin
